kmir/k-src/mir-rvalue.md

```diff
@@ -0,0 +1,371 @@
+  rule evalBinaryOpImpl(FN_KEY, Mul, X, Y)    => {evalOperand(FN_KEY, X)}:>Int *Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Div, X, Y)    => {evalOperand(FN_KEY, X)}:>Int /Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Rem, X, Y)    => {evalOperand(FN_KEY, X)}:>Int %Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, BitXor, X, Y) => {evalOperand(FN_KEY, X)}:>Int xorInt {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, BitOr, X, Y)  => {evalOperand(FN_KEY, X)}:>Int |Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, BitAnd, X, Y) => {evalOperand(FN_KEY, X)}:>Int &Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Shl, X, Y)    => {evalOperand(FN_KEY, X)}:>Int <<Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Shr, X, Y)    => {evalOperand(FN_KEY, X)}:>Int >>Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Eq, X, Y)     => {evalOperand(FN_KEY, X)}:>Int ==Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Lt, X, Y)     => {evalOperand(FN_KEY, X)}:>Int <Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Le, X, Y)     => {evalOperand(FN_KEY, X)}:>Int <=Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Ne, X, Y)     => {evalOperand(FN_KEY, X)}:>Int =/=Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Ge, X, Y)     => {evalOperand(FN_KEY, X)}:>Int >=Int {evalOperand(FN_KEY, Y)}:>Int
+  rule evalBinaryOpImpl(FN_KEY, Gt, X, Y)     => {evalOperand(FN_KEY, X)}:>Int >Int {evalOperand(FN_KEY, Y)}:>Int
+  // rule evalBinaryOpImpl("Offset", X, Y) => "not implemented"
+```
+
+### Constant evaluation.
+//TODO: implement other cases.
+
+```k
+  syntax MIRValue ::= evalConstantValue(ConstantValue) [function]
+  //-------------------------------------------------------------
+  rule evalConstantValue(VALUE:UnsignedLiteral) => UnsignedLiteral2Int(VALUE)
+  rule evalConstantValue(VALUE:SignedLiteral)   => SignedLiteral2Int(VALUE)
+  rule evalConstantValue(VALUE:StringLiteral)   => StringLitertal2String(VALUE)
+  rule evalConstantValue(( ))                   => Unit
+  rule evalConstantValue(VALUE:Bool)            => VALUE
+//  rule evalConstantValue(_VALUE)              => "Error: evalConstantValue --- unsupported ConstantValue" [owise]
+```
+
+### `Local` evaluation
+
+Locals only makes sense withing a function-like, hence we evaluate them as a contextual function that grabs the value from the function-like's environment:
+
+```k
+  syntax MIRValue ::= evalLocal(FunctionLikeKey, Local) [function]
+  //--------------------------------------------------------------
+  rule [[ evalLocal(FN_KEY, LOCAL) => VALUE ]]
+    <function>
+      <fnKey> FN_KEY </fnKey>
+      <localDecl>
+        <index> INDEX </index>
+        <value> VALUE </value>
+        ...
+      </localDecl>
+      ...
+    </function>
+    requires  INDEX ==Int Local2Int(LOCAL)
```

Comment:
I've lost track of the definition of `LOCAL`, except it is defined as `LOCAL = LocalToken`. Is it merely int values? 
How do we differenciate a concrete value and a place (memory) that hold a value? Or am I missing anything?

Reply:
The `Local` sort is indeed defined as `LocalToken`, and the latter is defined as:
```
  syntax LocalToken      ::= r"_[0-9]+"  [token(2)]
```

I.e. it's a "typed string" (see the relevant [section](https://github.com/runtimeverification/k/tree/master/k-distribution/k-tutorial/1_basic/03_parsing#tokens) of the K tutorial.

Now, the `<localDecl>` cell sort has it's first subcell `<index>` hold an `Int`. The side condition of the rule in questions requries the syntactic `Local` (for example, `_1`) to match the semantics `Int` (for example `1`).

Perhaps it makes sense to defined the `<index>` cell to hold `Local`, thus turning it into a typed string instead of an `Int`.